plugins/Goals/RecordBuilders/GeneralGoalsRecords.php

```diff
@@ -0,0 +1,234 @@
+<?php
+/**
+ * Matomo - free/libre analytics platform
+ *
+ * @link https://matomo.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ *
+ */
+
+namespace Piwik\Plugins\Goals\RecordBuilders;
+
+use Piwik\ArchiveProcessor;
+use Piwik\ArchiveProcessor\Record;
+use Piwik\DataAccess\LogAggregator;
+use Piwik\DataArray;
+use Piwik\DataTable;
+use Piwik\Metrics;
+use Piwik\Plugin\Manager;
+use Piwik\Plugins\Goals\API;
+use Piwik\Plugins\Goals\Archiver;
+use Piwik\Plugins\Goals\Goals;
+use Piwik\Tracker\GoalManager;
+
+class GeneralGoalsRecords extends Base
+{
+    const VISITS_UNTIL_RECORD_NAME = 'visits_until_conv';
+    const DAYS_UNTIL_CONV_RECORD_NAME = 'days_until_conv';
+    const VISITS_COUNT_FIELD = 'visitor_count_visits';
+    const LOG_CONVERSION_TABLE = 'log_conversion';
+    const SECONDS_SINCE_FIRST_VISIT_FIELD = 'visitor_seconds_since_first';
+
+    /**
+     * This array stores the ranges to use when displaying the 'visits to conversion' report
+     */
+    public static $visitCountRanges = array(
+        array(1, 1),
+        array(2, 2),
+        array(3, 3),
+        array(4, 4),
+        array(5, 5),
+        array(6, 6),
+        array(7, 7),
+        array(8, 8),
+        array(9, 14),
+        array(15, 25),
+        array(26, 50),
+        array(51, 100),
+        array(100)
+    );
+
+    /**
+     * This array stores the ranges to use when displaying the 'days to conversion' report
+     */
+    public static $daysToConvRanges = array(
+        array(0, 0),
+        array(1, 1),
+        array(2, 2),
+        array(3, 3),
+        array(4, 4),
+        array(5, 5),
+        array(6, 6),
+        array(7, 7),
+        array(8, 14),
+        array(15, 30),
+        array(31, 60),
+        array(61, 120),
+        array(121, 364),
+        array(364)
+    );
+
+    protected function aggregate(ArchiveProcessor $archiveProcessor)
```

Comment:
> Maybe it would also force people more to split queries up into multiple RecordBuilders and not just squeeze everything into one. We would be force to have one record builder per "raw data query".

I thought about something like this, but it's unfortunately not do-able. For some plugins it's one big log aggregation query for one record, but some plugins (like CustomReports or CrashAnalytics) will have to do multiple queries to build a single record.

We could maybe add some syntatic sugar to LogAggregator... something like:

```
$table = new DataTable();

$logAggregator
    ->query('log_crash_event')
    ->selectDimension(...)
    ->aggregateMetric(...)
    ...
    ->forEachRow(function ($row)  {
        // ... add to $table ...
    })
    // query for subtables
    ->query('log_crash_event')
    ...
    ->forEachRow(function ($row) {
        $label = $row['label'];
        $subtable = $table->getRowFromLabel($label)->getSubtable();
        $subtable->addRow(...);
    });

return [$recordname => $table];
```

then say each RecordBuilder should have as few log table queries as possible. Or I guess we could just say that w/o the rest of it. Not sure what the best or most idiomatic approach would be.

Reply:
Maybe we could have two options? Like in the base class we have a 

```php
public function aggregateDayReport(){
     $dataArray = $this->makeDataArray(); // or have something like makeEmptyRow () or ...
     $sql = $this->queryData();
     $query = $dbReader->query($sql['query'], $sql['bind'])
    while ($row = $query->fetch()) {
      $dataArray = $this->aggregateMetrics($row, $dataArray); // or maybe we can get rid of dataArray and have something like $this->sumRow();
   }
   return $this->buildDataTables($dataArray);
}
```

Then in a more complex case, they could overwrite `aggregateDayReport` (might be different name) and completely customise behaviour. Maybe in sometimes people wouldn't even need to overwrite `aggregateMetrics` or `buildDataTables()`. That's just a random thought. The actually needed may be different. I'm just meaning the concept. 

We are doing this for example with the `render` method in the `Report` base class. By default it works out of the box and in special cases you can overwrite it. 

Not sure what the actual "methods" would be in that case as I haven't looked across archivers what would make the most sense and make it most understandable for engineers etc. Just a thought. I think we should take this opportunity and make it more simple if we can.

Personally, I always found the concept of DataArray and DataTables confusing. Not sure if we could directly work with DataTables or just a plan array.

Just some thoughts.

Reply:
It would also make it more testable

Reply:
@tsteur 

> Maybe we could have two options? ... Then in a more complex case, they could overwrite aggregateDayReport (might be different name) and completely customise behaviour. Maybe in sometimes people wouldn't even need to overwrite aggregateMetrics or buildDataTables(). That's just a random thought. The actually needed may be different. I'm just meaning the concept.

Given how varied plugin archiving can be (eg, think Referrers vs. FormAnalytics vs. Actions vs. Goals) I'm not sure how useful it would be, or if it would even be possible to have some universal default implementation. Unless you can conceive of such an implementation that might be used by a majority of existing plugins? Personally I don't know what that would look like.

The only way I can see to make it simpler to write and understand log aggregation code, would be some kind of API that makes it clear the process is simply:

1. aggregate log data
2. for each aggregated log data, load into result record(s)
3. return result record(s)

Perhaps RecordBuilders don't need to be `@api` initially and perhaps a potentially simpler API will become clearer when more plugins are converted?

> Personally, I always found the concept of DataArray and DataTables confusing. Not sure if we could directly work with DataTables or just a plan array.

DataArrays are unnecessary, just an intermediate structure. DataTables can always be built directly, which is probably more memory efficient.